### ISSUE TITLE
FeatureFinderMetaboIdent issues

### ISSUE THREAD

**Issue by eeko-kon** [OpenMS/OpenMS#6058]
Before and after re-quantification the intensities often differ and after manual testing, FFM has a more reasonable result. I am working again with the benchmark data [MTBLS733](https://www.ebi.ac.uk/metabolights/MTBLS733/files) and I know exactly what the ratio of the metabolites should be across different samples, and it looks like FFM scores quite well on the concentration fold-changes (84% of the discriminating markers), whereas FFMident is off in a lot of cases (72% of the discriminating markers). Taking a closer look at the data, I noticed that FFMident sometimes traces the feature a lot later than FFM (when the intensity actually starts dropping).

This is one example, but I have investigated a few other features that were not detected with FFMident or that were detected similarly to the examples in the photo (1st image FFM and 2nd image FFMident):

<img width="1128" alt="mzML and FFM (sample1)" src="https://user-images.githubusercontent.com/54527247/164969616-c72311a3-6292-4b5f-9c37-e8884438882f.png">
<img width="1128" alt="mzML and FFM (sample1)" src="https://user-images.githubusercontent.com/54527247/164969623-a8f1e40a-d2d6-4ad0-ae0f-4d5d625aba35.png">

<img width="1120" alt="mzML and FFM (sample 4)" src="https://user-images.githubusercontent.com/54527247/164969667-941dea2e-f62f-4813-98cb-a204901820a2.png">
<img width="1119" alt="mzML and FFMident (sample 4)" src="https://user-images.githubusercontent.com/54527247/164969676-787910f3-a6eb-403e-bfa6-671106f3bb95.png">

Sometimes it starts detecting a feature even before the mass trace begins:
<img width="1117" alt="mzML and FFM (sample 3)" src="https://user-images.githubusercontent.com/54527247/164969770-16bd4e45-1d6a-4f90-b142-233b46755945.png">
<img width="1133" alt="mzML and FFMident (sample 3)" src="https://user-images.githubusercontent.com/54527247/164969775-f42a580c-d96f-470b-ada1-44bbd044c011.png">

Sometimes it will not detect it at all:
<img width="1054" alt="mzML and FFM (chlorine group) (sample 5)" src="https://user-images.githubusercontent.com/54527247/164969785-c8f6d288-2d7c-46e4-bb4c-8e9a88dcfc44.png">
<img width="1120" alt="mzML and FFMident (sample 5)" src="https://user-images.githubusercontent.com/54527247/164969793-bf7e3786-8ad2-40ec-bb5f-844a97f19661.png">

The mzML and featureXML files that are compared in TOPPView are the aligned ones.

**Comment by jpfeuffer** [comment]
Very nice find. I think that explains some discrepancies in our proteomics workflows as well.
We had similar experiences in low signal areas.
Regarding the "before the trace starts": it might be fitting a modified gaussian profile and cannot model it as steep as the data.
Since this uses swath algorithms for signal extraction and it is used in FFID and FFMID, I think it would be worthwhile to put some effort into investigating collectively.
@OpenMS/core

**Comment by jpfeuffer** [comment]
By the way, could it be that you used the same figure twice for the first example?

**Comment by eeko-kon** [comment]
Yes, you are right :) Just corrected it.

I agree that I notice those errors in low signal features that elute at a busy area of the chromatogram (first 30-50 seconds usually). 

Another outcome of this issue is probably that with this specific data, after FFMident, there are more missing values of true features than after FFM.

**Comment by hendrikweisser** [comment]
@eeko-kon: Thanks for reporting this. Could you post TOPPView screenshots of the extracted chromatograms for the features that FFMetId gets wrong (via parameter "chrom_out")?

**Comment by eeko-kon** [comment]
> @eeko-kon: Thanks for reporting this. Could you post TOPPView screenshots of the extracted chromatograms for the features that FFMetId gets wrong (via parameter "chrom_out")?

For EIC, should I use FileFilter?

**Comment by hendrikweisser** [comment]
> For EIC, should I use FileFilter?

I don't think I understand the question. If you make an "id" input file to FFMetId with just the entries you're interested in, you should get chromatograms for only those.

**Comment by eeko-kon** [comment]
Of course(!!), sorry Hendrik I use it for untargeted metabolomics and I just forgot its main function

**Comment by eeko-kon** [comment]
An example of completely missing the feature:
This is the aligned raw mzml vs the FFM XML (normal mass trace detection):
![Screenshot 2022-04-25 at 18 12 32](https://user-images.githubusercontent.com/54527247/165130027-1e64bfbf-78cc-4015-80c1-55a32ae26be4.png)

This is the aligned raw mzml vs FFMident XML and mzML:
![Screenshot 2022-04-25 at 18 12 53](https://user-images.githubusercontent.com/54527247/165130067-a5d8bd07-544a-4dbf-ae33-ff4c910f81e4.png)

**Comment by hendrikweisser** [comment]
When features are unexpectedly missing from the FFMetId results, this may be caused by overlap filtering. In order to avoid attributing the same signal intensities to different features, FFMetId detects overlaps between mass traces of features and removes overlapping ones (keeping only the one that's closest in RT to its target RT). This may happen e.g. in the last screenshot above, where the second mass trace from the lower target may overlap the first mass trace from the target above it.
Was there a log message about "resolving overlaps" during the FFMetId run?
Currently there's no way to opt out of overlap filtering, but we could easily add a flag for that.

**Comment by hendrikweisser** [comment]
@eeko-kon: For a representative feature with wrong RT range where you saved the extracted chromatograms, can you please post the TOPPView chromatogram view? It should look something like this:

![image](https://user-images.githubusercontent.com/2251068/165280533-e9fe85ff-42b4-49fb-87cf-ead6ba2258a3.png)

**Comment by eeko-kon** [comment]
> Was there a log message about "resolving overlaps" during the FFMetId run?

This is the only warning I am getting:

`WARNING in SignalToNoiseEstimatorMedian: 100% of all Signal-to-Noise estimates are too high, because the median was found in the rightmost histogram-bin. You should consider increasing 'max_intensity' (and maybe 'bin_count' with it, to keep bin width reasonable)> occurred 5 times`

**Comment by hendrikweisser** [comment]
> This is the only warning I am getting:

Not a warning, just a message saying: "... features left after resolving overlaps (involving ... features in ... groups)."

**Comment by eeko-kon** [comment]
Ah. Still no, nothing involving overlaps.

**Comment by hendrikweisser** [comment]
> Ah. Still no, nothing involving overlaps.

:-( I hoped that would be the explanation for missing features.
One more thing you can try: Use the "candidates_out" parameter to generate a featureXML file with all the candidate features that were originally detected in the chromatograms. If you get no candidate where you expect it, the problem is in the OpenSWATH chromatogram feature detection. If you get a candidate but no "final" feature, it's a problem with the selection/filtering.

**Comment by eeko-kon** [comment]
> Ah. Still no, nothing involving overlaps.

I ran more files and got it!

`811 features left after resolving overlaps (involving 40 features in 20 groups).`

**Comment by eeko-kon** [comment]
> @eeko-kon: For a representative feature with wrong RT range where you saved the extracted chromatograms, can you please post the TOPPView chromatogram view? It should look something like this:

Here you are:

FFM chrom_out:
![Screenshot 2022-04-26 at 14 56 36](https://user-images.githubusercontent.com/54527247/165304880-aa28dd7a-25b0-4404-9c7f-e6732acdd383.png)

FFMident chrom_out:

![Screenshot 2022-04-26 at 14 52 36](https://user-images.githubusercontent.com/54527247/165304680-c39c6c00-5480-4690-9fb1-4f95c6919470.png)

**Comment by hendrikweisser** [comment]
> Is this what you meant?

Yes! You can see in the second image that the mass trace didn't get extracted properly by FFMetId - which then makes it impossible to detect a feature. Where does the target mass come from in this case? What m/z tolerance do you use for chromatogram extraction (parameter "extract:mz_window")? Can you make sure that the mass value isn't off slightly, and/or increase the tolerance before running FFMetId again?

**Comment by timosachsenberg** [comment]
I wonder if we could do better here by default? e.g. estimate that tolerance window somehow or detect mass drifts in certain samples.

**Comment by jpfeuffer** [comment]
FFM calculates the std dev of the mass traces anyway. Why not use it for the tolerance in requantification?

**Comment by timosachsenberg** [comment]
- I think it is not a true stdev but maybe can be used
- This would not capture calibration error

is this based on a consensus feature? Because then we could look at the stdev in the consensus

**Comment by axelwalter** [comment]
Yes, it is based on the consensus feature. Could we just take the highest stdev in the consensus features and use that as the mz_window?

**Comment by timosachsenberg** [comment]
For a very rough estimate, I think this could help, but my feeling is that with so many samples choosing e.g, 3*stdev will not give a lot of improvement e.g. easily extract to many.
I think we need to break this down into the types of errors.

The variance in each mass trace tells us something about the resolution and e.g., intensity dependent mass error.
The variance in a consensus feature tells us probably not much if we look at an individual consensus feature (just too much randomness and no clear reason why variance depends on the actual analyte). We could look at a global variance in consensus features (maybe using MAD or something more robust) but this will only give a rough estimate.

Just an idea: can we use the m/zs of features for requantification instead of the consensus m/z? This could compensate the run specific mass errors. This should increase the number of features that have been found by FFId close to the number of FFM if the feature was already found in FFM.

We would then still need to solve the problem of requantifying features that are missing in a run (e.g., not found by FFM). Here we would probably need to check if we can choose a smarter extraction position than the consensus m/z (e.g., we could try to calculate a mass offset that is specific for that run (e.g., a global mass error that we could add to the consensus m/z or even an RT dependent mass error for that run - by looking at the other feature identified in that run)

**Comment by eeko-kon** [comment]
> Just an idea: can we use the m/zs of features for requantification instead of the consensus m/z? This could compensate the run specific mass errors. This should increase the number of features that have been found by FFId close to the number of FFM if the feature was already found in FFM.

Could we take the feature with the highest quality from each consensus?

**Comment by timosachsenberg** [comment]
Wouldn't it be better to use the one that FFM found in exactly that file?
e.g., if features are off +-10 ppm in the consensus then the high quality feature could be at -10 ppm and we use that to extract a feature which would be at +10 ppm which will likely miss it.

**Comment by eeko-kon** [comment]
Hm ok I see. But then if we do not collapse the features, wouldn't FFMident find overlapping ones all the time? Since there would be a RT/mz deviation in every file. Or how are you thinking of proceeding with identical features across diff files?

**Comment by axelwalter** [comment]
We do FFMID for requantification of missing features, so often there is no feature from FFM for that sample from which we can take the m/z.

**Comment by timosachsenberg** [comment]
Yes this was I meant with "We would then still need to solve the problem of requantifying features that are missing in a run (e.g., not found by FFM). Here we would probably need to check if we can choose a smarter extraction position than the consensus m/z (e.g., we could try to calculate a mass offset that is specific for that run (e.g., a global mass error that we could add to the consensus m/z or even an RT dependent mass error for that run - by looking at the other feature identified in that run)".
But don't you also do requantification with FFMid of feature found in FFM? I would guess so

**Comment by eeko-kon** [comment]
> We do FFMID for requantification of missing features, so often there is no feature from FFM for that sample from which we can take the m/z.

In the consensusXML after FFM there is always the mz and RT that we will later re-quantify (always a feature present in one sample) but with missing values across other samples, so we essentially do "gap filling"

**Comment by eeko-kon** [comment]
But you are right that if we extract features from FFM and use those for the re-quantification of the respective file, then we will end up with max the same number of features per file or even less

**Comment by axelwalter** [comment]
Yes, exactly. So I like the suggestion to calculate the global mass error for the specific run and use that for FFMID.

**Comment by eeko-kon** [comment]
I re-calculated everything and here are the results that I generate with the following parameters:

**FFM**: 

`FeatureFinderMetabo -in {input} -out {output.out} -algorithm:common:noise_threshold_int "1.0e03" -algorithm:mtd:mass_error_ppm "10.0" -algorithm:epd:width_filtering "fixed" -algorithm:ffm:isotope_filtering_model "none" -algorithm:ffm:remove_single_traces "false" -algorithm:ffm:report_convex_hulls "true"`

- The consensus file has 31,345 "high-quality" features (features with no NaN values across replicates), 771 annotated truth features (92.2%)

Quantification calculations:
- 766 annotated ground truth features without any nan values - 744 accurate quantification predictions (97.12%) 

**FFMid**: 

`FeatureFinderMetaboIdent -id {input.var1} -in {input.var2} -out {output.out1} -extract:mz_window 10.0 -extract:rt_window 30.0 -detect:peak_width 60.0 `

- consensus has 73,275 features, 774  annotated truth features (92.5%)

Quantification calculations:
- 759 annotated ground truth features without any nan values - 737 accurate quantification predictions (97.10%) 

**The paper takes into account only approximately 70 compounds for the quantification (discrimination markers) which lowers the percentage of accurate quantification to 85.3% for FFM and 84.6% for FFMident for the publication**

If you have time to take a look, I would appreciate it! I am getting a little tired with all this data and I'm afraid that I might make mistakes. This is the link to all files: 
https://drive.google.com/drive/folders/1KPOPV6TjEy9qfziG7C_I3VDb-Yf1e-9u?usp=sharing

However, there are in total **143 ground truth features in FFMident that have one or more missing values across the samples**. However, in FFM **none** of them have any NaN values (across all samples/replicates). Nevertheless, the intensities -when measured- are quite precise now (roughly the same between FFM and FFMident).

I will randomly investigate some of those missing values to try and find a pattern. 

@axel:

Summary: There is an almost identical number of missing values in FFM and FFMident with lower int threshold in FFM:

1) Int thr 10^3 (FFM):

total number of features in consensus from FFM: 85,286

SA1 sum of NaN values 47073
SA2 sum of NaN values 49028
SA3 sum of NaN values 47923
SA4 sum of NaN values 48072
SA5 sum of NaN values 48345
SB1 sum of NaN values 48104
SB2 sum of NaN values 48557
SB3 sum of NaN values 47925
SB4 sum of NaN values 48435
SB5 sum of NaN values 47964

total number of features in consensus from FFMident: 73,275

SA1 sum of NaN values 25142
SA2 sum of NaN values 22627
SA3 sum of NaN values 21752
SA4 sum of NaN values 21679
SA5 sum of NaN values 21426
SB1 sum of NaN values 20888
SB2 sum of NaN values 21139
SB3 sum of NaN values 22295
SB4 sum of NaN values 24612
SB5 sum of NaN values 26549

2) Int thr 100:

total number of features in consensus from FFM: 85,288

SA1 sum of NaN values 47073
SA2 sum of NaN values 49029
SA3 sum of NaN values 47923
SA4 sum of NaN values 48071
SA5 sum of NaN values 48346
SB1 sum of NaN values 48105
SB2 sum of NaN values 48557
SB3 sum of NaN values 47924
SB4 sum of NaN values 48432
SB5 sum of NaN values 47961

total number of features in consensus from FFMident: 73,289

SA1 sum of NaN values 25162
SA2 sum of NaN values 22636
SA3 sum of NaN values 21765
SA4 sum of NaN values 21694
SA5 sum of NaN values 21436
SB1 sum of NaN values 20896
SB2 sum of NaN values 21145
SB3 sum of NaN values 22321
SB4 sum of NaN values 24618
SB5 sum of NaN values 26547

Re-quantification is obviously necessary, but I think the best way to go (at least for the near future) is to use FFMident for gap-filling instead of total re-quantification. Meaning that we should fill only the missing values from the consensus table generated from FFM with values extracted from FFMident. Currently I will do that via merging the tables (small python script), but if we cannot find a fix it would be a good idea in general to do that instead (computationally way less expensive also).

Let me know what you think - I will follow up with some TOPPView screenshots- be prepared.
[missing_values_FFMident.csv](https://github.com/OpenMS/OpenMS/files/8577087/missing_values_FFMident.csv)
[FFM_no_NaN.csv](https://github.com/OpenMS/OpenMS/files/8577215/FFM_no_NaN.csv)

**Comment by jpfeuffer** [comment]
By the way, my experience with the dataset was that it also has very short features in the beginning (2-3 seconds fwhm?). But you have the default of 1s there, right? 

Also did you include Chloride in the elements for isotope distance and monoisotope calculation?

**Comment by eeko-kon** [comment]
> By the way, my experience with the dataset was that it also has very short features in the beginning (2-3 seconds fwhm?). But you have the default of 1s there, right? Maybe Lower min_trace_length.

I will give it a try (running it now). I have fwhm on default, yes. 

> Also did you include Chloride in the elements for isotope distance and monoisotope calculation?

I am not scoring by elements, I found it a bit inaccurate when I was using my own data. I will give a try too.

Here are 4 different features that did not get picked up from FFMident (TOPPView investigation)
[Feature_investigation.docx](https://github.com/OpenMS/OpenMS/files/8577974/Feature_investigation.docx)

**Comment by timosachsenberg** [comment]
Another thing we could investigate:
Use the peak fitting in https://pubs.acs.org/doi/full/10.1021/acs.analchem.0c03421 to complete cut peaks or saturated peaks

**Comment by eeko-kon** [comment]
This FFMident issue is "fixed" with the following strategy:

- Split the consensusXML to features that have no missing values and features that have at least one NaN value in all samples
- Re-load the former to featureXMLs
- Use FFMident to **only re-quantify features that have missing values** in the samples
- Merge the "complete" with the FFMident re-quantified featureXMLs for every sample
- Create a consensusXML that has fewer NaN values& more accurate quantifications

This is all currently done using pyOpenMS thanks to @axelwalter, but I think it would be a great advantage to have it as an untargeted metabolomics re-quantification TOPP tool.

**Comment by jcharkow** [comment]
@alhigaylan is this at all related to the issue you are having?
